addons/mass_mailing/models/mailing.py

```diff
@@ -518,7 +547,15 @@ def action_retry_failed(self):
         ])
         failed_mails.mapped('mailing_trace_ids').unlink()
         failed_mails.unlink()
-        self.write({'state': 'in_queue'})
+        mailing_send_immediate_nbr = self._get_mass_mailing_immediate_send_nbr()
+        if mailing_send_immediate_nbr <= 0:
+            self.state = 'in_queue'
+        else:
+            mailings_to_send_immediately = self.filtered(lambda mailing: len(mailing._get_remaining_recipients()) <= mailing_send_immediate_nbr)
+            mailings_to_send_immediately.state = 'sending'
+            mailings_to_send_immediately.action_send_mail()
+            (self - mailings_to_send_immediately).state = 'in_queue'
```

Comment:
all this code is almost the same as `action_launch`, can we make a method `_send_or_put_in_queue` which take in argument the number of recipients ?

What do you think ?